### PR TITLE
Deprecated reduce in order to be symmetric to the fold API

### PR DESCRIPTION
**Pull request by danieldietrich**
This PR deprecates `IterableOnce.reduce` in order to be 'in sync' with `fold`.

See [this discussion](https://twitter.com/julienrf/status/1087663655200669701).

### CODE REVIEW

**Review by julienrf — APPROVED** [review]

**Comment by ashawley** [comment]
Is removing `scan`, `fold` and `reduce` an optimization issue?  Aren't they there for compatibility with non-sequential types like `Option` or `Either`?   The arbitrary direction of these functions that the implementation determines have a long legacy in functional programming.  Seems odd to remove them.

**Comment by NthPortal** [comment]
@ashawley I believe those methods were meant to be a super-interface for parallel and non-parallel collections, because `<op>{Left,Right}` can't be done in parallel, but `<op>` is not specified to have an order/direction and can.

However, now that the parallel collections are their own library and framework and not part of the regular collection framework, the `<op>` methods aren't useful over `<op>Left`.

**Comment by julienrf** [comment]
@ashawley These operations have no utility on sequential collections. They are defined as an alias to their explicit left-to-right analogous.

**Comment by ashawley** [comment]
> These operations have no utility on sequential collections.

Agreed, that's why I mentioned compatibility with non-sequential collections.  This doesn't seem like a very holistic change, even though it seems like an obvious local optimization in the context of collections.

> They are defined as an alias to their explicit left-to-right analogous.

Sure, I understand they are aliases in the collections library and violate the dry principle, just seems strange to remove them for that reason alone.

> I believe those methods were meant to be a super-interface for parallel and non-parallel collections

Yes, parallel collections are another example of why we'd want to keep them.  

PS.  I also always liked having them around for when I didn't need to care about the order of operation for either correctness or performance reasons, but the standard library authors would make sure to do the right thing™.

**Comment by SethTisue** [comment]
if this is going to be deprecated, shouldn't it happen in scala/scala rather than here, given our promises around Scala 2.14 & 3 sharing the same stdlib?

**Comment by SethTisue** [comment]
@axel22 opinion?

**Comment by axel22** [comment]
I don't have a very strong opinion here, and I am sympathetic towards both the arguments for and against keeping `fold`/`reduce`.

As a minor remark, from an intuition standpoint, it seems to me that having a method like `foldLeft` or `foldRight` on non-sequences and non-ordered collections (i.e. non-linked-maps/sets and non-tree-maps/sets) might have been a counterintuitive choice to begin with. IMO, it would have been better to have only `fold`/`reduce` on traversable and iterable collections, with `<op>Left`/`<op>Right` only for collections that actually have a notion of order.
But since these decisions were made a long time ago, it's too costly to change them now.

**Comment by joroKr21** [comment]
The problem with fold and reduce for unordered collections is that they are not powerful enough (due to the subtyping constraints). You would rather have the full generality of foldMap or a fold with three arguments but that may not be feasible for the stdlib.

**Comment by axel22** [comment]
Yes - in an ideal world, I'd say it makes sense to have `fold`/`reduce` with more general signatures (i.e. `def fold[B](z: B)(op: (B, A) => B): B`), and have a separate method (e.g. `foldUp` or something like that) for parallel reduction, which has a more restrictive interface.

**Comment by axel22** [comment]
Also - would you ever want to call `foldRight` on a `HashMap` or a `BinaryHeap`, or some other unordered collection? This does not seem very useful, since there is no guaranteed order anyway.

**Comment by ashawley** [comment]
Rather than derail this PR any further, I've created a ticket to continue the discussion.  This change, to deprecate `reduce`, should be merged for consistency with the other deprecations.  There's no reason to hold this up.  My proposal is to preserve all the methods not just `reduce`, anyway.

**Comment by szeiger** [comment]
Well, if we're going to preserve all methods (which I think is a good idea), deprecating even more is counter-productive and we should rather undeprecate `fold`. We can still decide to go one way or the other in 2.14.

**Comment by axel22** [comment]
I agree that maintaining compatibility with the old API has a high priority.

**Comment by SethTisue** [comment]
The CI failure is a missing checkfile update:

```
!![0m 1634 - run/t8428.scala                           [output differs]
% [0mdiff /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/run/t8428.check /home/jenkins/workspace/scala-2.13.x-validate-main@2/test/files/run/t8428-run.log--- empty
+++ t8428-run.log
@@ -1,0 +1,1 @@
+warning: there was one deprecation warning (since 2.13.0); re-run with -deprecation for details
```

discussion is ongoing at https://github.com/scala/bug/issues/11382

**Comment by danieldietrich** [comment]
@SethTisue thank you, I will take a look at the deprecation warning later today and update the PR accordingly.

**Comment by danieldietrich** [comment]
How do I make the 'combined' test green? Do I need to squash my commits and push-force? Or is it ok as-is because all CI tests are green?
Thank you for your help!

**Comment by ashawley** [comment]
You'll need to squash your commits together, which is the policy of this repo, anyway.

**Comment by danieldietrich** [comment]
@ashawley thank you for clarification!

**Comment by Ichoran** [comment]
@danieldietrich @julienrf - There is an important reason not to remove `fold`--it has a different (simpler) signature than `foldLeft` or `foldRight`.  If you don't know which way you're going, the natural form for the operation is `(A, A) => A`, namely with the same types.  And that is, in fact, how it is now.

I think that this simple variant should be present, as it is now, to document that it really doesn't matter how the fold happens.

And because `fold` has this useful variant, I think `reduce` should remain also, in short and simple form, for those cases where it doesn't matter.

You can always pick left or right if you care, but most of the time you won't.  (Note that `reduceLeft` and `reduceRight` already have to limit themselves to `(A, A) => A` and not `(B, A) => B` or `(A, B) => B`.)

**Comment by danieldietrich** [comment]
@Ichoran it is perfectly ok! This PR had only one goal: to restore consistency regarding the deprecation of similar methods. Meanwhile it turned out that the deprecation of fold should be reverted instead of deprecating reduce.

A positive result of this PR was that this topic has been discussed. I think the committers found a general consensus while taking all variables into account (ordered vs. unordered collections vs. backw'compat, usefulness, ...).

I suggest that I close this PR and the main committers take care of this issue either in 2.13-RC* or 2.14.
Ok?

**Comment by SethTisue** [comment]
okay, let's undeprecate `fold` then, volunteer for that?

**Comment by danieldietrich** [comment]
@SethTisue Here 👋🏼, I will do it.

Only `fold` or also `scan`? Please see https://github.com/scala/scala/pull/6866

**Comment by julienrf** [comment]
`fold` **and** `scan`, indeed.

**Comment by danieldietrich** [comment]
Thank you, I will do that later today.

**Pull request closed without merging**
